android/engine/src/main/java/org/smartregister/fhircore/engine/task/WorkflowCarePlanGenerator.kt

```diff
@@ -0,0 +1,329 @@
+    proposedCarePlan: CarePlan,
+    carePlanOfRecord: CarePlan,
+  ) {
+    val resourceList = createProposedRequestResources(proposedCarePlan.contained)
+    updateCarePlanWithProtocol(carePlanOfRecord, proposedCarePlan.instantiatesCanonical)
+    addRequestResourcesToCarePlanOfRecord(carePlanOfRecord, resourceList)
+
+    fhirEngine.update(carePlanOfRecord)
+    linkRequestResourcesToCarePlan(carePlanOfRecord, resourceList)
+  }
+
+  /** Update status of a [CarePlan] activity */
+  private suspend fun updateCarePlanStatus(
+    carePlan: CarePlan,
+    requestedActivityResource: Resource,
+    carePlanActivityStatus: CarePlan.CarePlanActivityStatus,
+    outcomeReferences: List<Reference>,
+  ) {
+    if (carePlan.isEmpty) return
+    for (activity in carePlan.activity) {
+      if (
+        activity.reference.reference.equals(
+          requestedActivityResource.fhirType() + "/" + IdType(requestedActivityResource.id).idPart,
+        )
+      ) {
+        activity.detail.status = carePlanActivityStatus
+        activity.outcomeReference = outcomeReferences
+        fhirEngine.update(carePlan)
+        break
+      }
+    }
+  }
+
+  /**
+   * Find and update the status of the [CarePlan] activity as per the corresponding request resource
+   * status
+   */
+  suspend fun updateCarePlanActivity(
+    requestResource: Resource,
+    requestResourceStatus: String,
+    outcomeReferences: List<Reference>,
+    updateCarePlan: Boolean = true,
+  ) {
+    val carePlanActivityStatus: CarePlan.CarePlanActivityStatus
+    val carePlan: CarePlan
+    when (requestResource.fhirType()) {
+      "Task" -> {
+        updateRequestResourceStatus(requestResource as Task, requestResourceStatus)
+        if (updateCarePlan) {
+          carePlanActivityStatus = mapRequestResourceStatusToCarePlanStatus(requestResource)
+          carePlan =
+            if (requestResource.hasBasedOn()) {
+              fhirEngine.get(
+                ResourceType.CarePlan,
+                IdType(requestResource.basedOnFirstRep.referenceElement.value).idPart,
+              ) as CarePlan
+            } else {
+              return
+            }
+          updateCarePlanStatus(carePlan, requestResource, carePlanActivityStatus, outcomeReferences)
+        }
+      }
+      "ServiceRequest" -> TODO("Not supported yet")
+      "MedicationRequest" -> TODO("Not supported yet")
+      "SupplyRequest" -> TODO("Not supported yet")
+      "Procedure" -> TODO("Not supported yet")
+      "DiagnosticReport" -> TODO("Not supported yet")
+      "Communication" -> TODO("Not supported yet")
+      "CommunicationRequest" -> TODO("Not supported yet")
+      "RequestGroup" -> {}
+      else -> TODO("Not a valid request resource")
+    }
+  }
+
+  /** Update [Task] status */
+  suspend fun updateRequestResourceStatus(resource: Task, status: String) {
+    resource.status = Task.TaskStatus.valueOf(status)
+    resource.executionPeriod.end = Date.from(Instant.now())
```

Comment:
whether Task ends or not depends on task status. For our current flow it might be fine but task statuses could intermediary or on-hold as well which means that pause it for now but it is not ended can be resumed later on

Reply:
@owais-vd could we reuse the already existing Task status update functionality?  That way this is handled in the same place throughout the app